src/DbPostgreSQLAdapter/Record.cpp

```diff
@@ -0,0 +1,114 @@
+#include "stdafx.h"
+#include "Record.h"
+
+#include "Field.h"
+#include "FieldValue.h"
+
+#include "DbAdapterInterface/IRecordSet.h"
+
+#include "libpq-fe.h"
+
+namespace {
+	std::chrono::system_clock::time_point stringISOToDateTime(const std::string& dateTime)
+	{
+		std::chrono::system_clock::time_point timePointDateTime;
+		std::istringstream{ dateTime } >> std::chrono::parse("%F %T%z", timePointDateTime);
+		return timePointDateTime;
+	}
+}
+
+namespace systelab::db::postgresql {
+
+	Record::Record(IRecordSet& recordSet, const PGresult* statementResult, const int rowIndex)
+	{
+		const unsigned int fieldsCount = recordSet.getFieldsCount();
+		for (unsigned int i = 0; i < fieldsCount; i++)
+		{
+			std::unique_ptr<IFieldValue> fieldValue;
+			const IField& field = recordSet.getField(i);
+			const unsigned int fieldIndex = field.getIndex();
+			if (PQgetisnull(statementResult, rowIndex, fieldIndex) == 1)
+			{
+				fieldValue.reset( new FieldValue(field));
+			}
+			else
+			{
+				FieldTypes fieldType = field.getType();
+				const std::string value = PQgetvalue(statementResult, rowIndex, fieldIndex);
+				switch(fieldType)
+				{
+					case BOOLEAN:
+						fieldValue = std::make_unique<FieldValue>(field, (value == "t"));
+						break;
+					case INT:
+						fieldValue = std::make_unique<FieldValue>(field, std::stoi(value));
+						break;
+					case DOUBLE:
+						fieldValue = std::make_unique<FieldValue>(field, std::stod(value));
+						break;
+					case STRING:
+						fieldValue = std::make_unique<FieldValue>(field, std::string(value));
+						break;
+
+					case DATETIME:
+						fieldValue = std::make_unique<FieldValue>(field, stringISOToDateTime(value));
+						break;
+					case BINARY:
+					default:
+						throw std::runtime_error( "Unknown field type." );
+				}
+			}
+
+			m_fieldValues.push_back(std::move(fieldValue));
+		}
+	}
+
+	Record::Record(std::vector<std::unique_ptr<IFieldValue>>& fieldValues)
+	{
+		const unsigned int nFieldValues = static_cast<unsigned int>(fieldValues.size());
+		for (unsigned int i = 0; i < nFieldValues; i++)
+		{
+			m_fieldValues.push_back(std::move(fieldValues.at(i)));
+		}
+	}
+
+	unsigned int Record::getFieldValuesCount() const
+	{
+		return static_cast<unsigned int>(m_fieldValues.size());
+	}
+
+	IFieldValue& Record::getFieldValue(unsigned int index) const
+	{
+		if (index >= m_fieldValues.size())
+		{
+			throw std::runtime_error("Invalid field value index");
+		}
+
+		return *(m_fieldValues.at(index).get());
```

Comment:
The `.at()` already throws a specific exception if out of bounds, so you can go straight and take advantage of it

Reply:
Done